src/main/java/org/broadinstitute/hellbender/utils/read/AlignmentUtils.java

```diff
@@ -200,48 +201,69 @@ public static GATKRead createReadAlignedToRef(final GATKRead originalRead,
+     * will be filled in with GAP_CHARACTER values for the read bases and 0's for base qualities to indicate that they don't exist.
+     *
+     * If the cigar for input read is all matches to the reference then this method will return references to the original
+     * read base/base quality byte arrays in the underlying SamRecord in order to save on array allocation/copying performance effects.
+     *
+     * @param read a read to return aligned to the reference
+     * @return A tuple of byte arrays where the first array corresponds to the bases aligned to the reference and second
+     *         array corresponds to the baseQualities aligned to the reference.
+     */
+    public static Pair<byte[], byte[]> getBasesAndBaseQualitiesAlignedOneToOne(final GATKRead read) {
+        return getBasesAndBaseQualitiesAlignedOneToOne(read, GAP_CHARACTER, (byte)0);
     }
 
-    public static byte[] getSequenceAlignedOneToOne(final GATKRead read, final Function<GATKRead, byte[]> bytesProvider, final byte padWith) {
+    private static Pair<byte[], byte[]> getBasesAndBaseQualitiesAlignedOneToOne(final GATKRead read, final byte gapCharacter, final byte qualityPadCharacter) {
         Utils.nonNull(read);
-        Utils.nonNull(bytesProvider);
-        final Cigar cigar = read.getCigar();
-        final byte[] sequence = bytesProvider.apply(read);
-
-        if (!cigar.containsOperator(CigarOperator.DELETION) && !cigar.containsOperator(CigarOperator.INSERTION)) {
-            return sequence;
+        final byte[] bases = read.getBasesNoCopy();
+        final byte[] baseQualities = read.getBaseQualitiesNoCopy();
+        final int numCigarElements = read.numCigarElements();
+        boolean sawIndel = false;
+
+        // Check if the cigar contains indels
+        for (int i = 0; i < numCigarElements; i++) {
+            final CigarOperator e = read.getCigarElement(i).getOperator();
+            if (e == CigarOperator.INSERTION || e == CigarOperator.DELETION) {
+                sawIndel = true;
+                break;
+            }
+        }
+        if (!sawIndel) {
+            return new ImmutablePair<>(bases, baseQualities);
         }
         else {
-            final byte[] paddedBases = new byte[CigarUtils.countRefBasesIncludingSoftClips(read, 0, cigar.numCigarElements())];
+            int numberRefBasesIncludingSoftclips = CigarUtils.countRefBasesIncludingSoftClips(read, 0, numCigarElements);
+            final byte[] paddedBases = new byte[numberRefBasesIncludingSoftclips];
+            final byte[] paddedBaseQualities = new byte[numberRefBasesIncludingSoftclips];
             int literalPos = 0;
             int paddedPos = 0;
-            for ( int i = 0; i < cigar.numCigarElements(); i++ ) {
-                final CigarElement ce = cigar.getCigarElement(i);
+            for ( int i = 0; i < read.numCigarElements(); i++ ) {
+                final CigarElement ce = read.getCigarElement(i);
```

Comment:
Use the already-initialized `numCigarElements` in the for loop condition here instead of `read.numCigarElements()` (another comment not addressed from last time)